### PR TITLE
[TF:MLIR] Legalize tf.diag with tf2xla

### PR DESCRIPTION
**Pull request by WindQAQ**
Legalize `tf.diag` with tf2xla as `MlirHloBuilder::Iota` is available.

### CODE REVIEW

**Comment by joker-eph** [comment]
Can you add an IR test under tensorflow/compiler/mlir/test?
I'm also curious if this pattern is complex or if it can be supported by a native MLIR lowering.

**Comment by joker-eph** [comment]
Test is here FYI: https://github.com/tensorflow/tensorflow/tree/master/tensorflow/compiler/mlir/xla/tests/legalize-tf-with-tf2xla.mlir

File: tensorflow/compiler/mlir/tensorflow/ir/tf_generated_ops.td [hunk 1]
```diff
@@ -2461,6 +2461,40 @@ this op runs. The length of the list is returned in two cases:
   );
 }
 
+def TF_DiagOp : TF_Op<"Diag", [NoSideEffect]> {
```
Comment (andyly) [inline tensorflow/compiler/mlir/tensorflow/ir/tf_generated_ops.td:2464]: Can you add `SameOperandsAndResultElementType` trait to this op?

Comment (WindQAQ) [reply tensorflow/compiler/mlir/tensorflow/ir/tf_generated_ops.td]: Thanks for pointing out this! Updated.

**Comment by WindQAQ** [comment]
Hi @joker-eph, I've updated the test. I'm also curious if there is any standard when talking about complicated IR lowering (maybe in terms of number of IRs generated)? Like in this case, the legalized IR is composed of less than 10 mhlo ops, is that complex or not? Thanks!

**Comment by smit-hinsu** [comment]
There isn't any objective standard to describe complexity of a legalization pattern. It could be the number of ops generated, lines of code for the pattern and test etc.

If you could implement rewrite patterns with reasonable effort natively without using the fallback kernel, that should be preferred.

**Review by smit-hinsu — DISMISSED** [review]

**Review by andyly — DISMISSED** [review]

File: tensorflow/compiler/tests/unary_ops_test.py [hunk 2]
```diff
@@ -95,8 +95,6 @@ def AssertAllEqual(self, result, expected, rtol, atol):
     """Tests that result and expeted are exactly equal."""
     self.assertAllEqual(result, expected)
 
-  @test_util.disable_mlir_bridge(
```
Comment (smit-hinsu) [inline tensorflow/compiler/tests/unary_ops_test.py:98]: This test has other failures as well. Keep this test disabled and change the description to "Handle complex element types in DiagPart op lowering"

**Review by smit-hinsu — DISMISSED** [review]

**Comment by gbaned** [comment]
@WindQAQ  Can you please fix build failures ? Thanks!

**Review by smit-hinsu — DISMISSED** [review]

**Review by andyly — DISMISSED** [review]

File: tensorflow/compiler/mlir/xla/tests/legalize-tf-with-tf2xla.mlir [hunk 3]
```diff
@@ -283,6 +283,20 @@ func @bessel_i1e(%arg0: tensor<3xf16>, %arg1: tensor<3xf32>, %arg2: tensor<3xf64
...
 }
 
+// CHECK-LABEL: diag
+func @diag(%arg0: tensor<2xf32>) -> tensor<2x2xf32> {
+  // CHECK: %[[ZERO:.*]]  = mhlo.constant dense<0.000000e+00> : tensor<2x2xf32>
+  // CHECK: %[[IOTA:.*]]  = "mhlo.iota"() {iota_dimension = 0 : i64} : () -> tensor<2xi32>
+  // CHECK: %[[BROADCAST1:.*]] = "mhlo.broadcast_in_dim"(%[[IOTA]]) {broadcast_dimensions = dense<1> : tensor<1xi64>} : (tensor<2xi32>) -> tensor<2x2xi32>
+  // CHECK: %[[BROADCAST0:.*]] = "mhlo.broadcast_in_dim"(%[[IOTA]]) {broadcast_dimensions = dense<0> : tensor<1xi64>} : (tensor<2xi32>) -> tensor<2x2xi32>
+  // CHECK: %[[EQ:.*]] = "mhlo.compare"(%[[BROADCAST1]], %[[BROADCAST0]]) {comparison_direction = "EQ"} : (tensor<2x2xi32>, tensor<2x2xi32>) -> tensor<2x2xi1>
+  // CHECK: %[[BROADCAST2:.*]] = "mhlo.broadcast_in_dim"(%[[ARG0]]) {broadcast_dimensions = dense<1> : tensor<1xi64>} : (tensor<2xf32>) -> tensor<2x2xf32>
```
Comment (smit-hinsu) [inline tensorflow/compiler/mlir/xla/tests/legalize-tf-with-tf2xla.mlir:293]: ARG0 is not defined. This commit doesn't implement lowering so just having a CHECK-NOT to make sure tf.Diag op is not present after the pass is enough.

Comment (WindQAQ) [reply tensorflow/compiler/mlir/xla/tests/legalize-tf-with-tf2xla.mlir]: Thanks you for clarification. Done!

**Review by smit-hinsu — APPROVED** [review]

**Pull request merged**